### PR TITLE
feat: create types for maps with constraints

### PR DESCRIPTION
**Pull request by barnabasJ**
### Contributor checklist

- [ ] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

### CODE REVIEW

File: lib/ash_graphql.ex [hunk 1]
```diff
@@ -398,6 +401,13 @@ defmodule AshGraphql do
         attribute_type = unwrap_type(attribute.type)
 
         case attribute_type do
+          Ash.Type.Map ->
+            if attribute.constraints[:fields] do
```
Comment (zachdaniel) [inline lib/ash_graphql.ex:405]: You'll probably need to map over the sub-fields the way that the `union` type does.

File: lib/resource/resource.ex [hunk 2]
```diff
@@ -2334,6 +2336,107 @@ defmodule AshGraphql.Resource do
     String.to_atom(to_string(type) <> "_sort_field")
   end
 
+  def map_definitions(resource, schema, env) do
+    resource = Ash.Type.NewType.subtype_of(resource)
```
Comment (zachdaniel) [inline lib/resource/resource.ex:2340]: I don't think we should need to do this.

File: lib/resource/resource.ex [hunk 3]
```diff
@@ -2334,6 +2336,107 @@ defmodule AshGraphql.Resource do
...
+            end
+          else
+            nil
+          end
+
+        [
+          type_name
+        ]
+        |> Enum.filter(& &1)
+        |> Enum.map(fn type_name ->
```
Comment (zachdaniel) [inline lib/resource/resource.ex:2392]: We need to do this differently for input and non-input, since some types have different input representations.

**Review by zachdaniel — DISMISSED** [review]
Looking great! Got some changes to make, and then ideally we can support this in nested cases.

**Comment by barnabasJ** [comment]
Nested maps should now work too.

**Review by zachdaniel — APPROVED** [review]

**Pull request merged**